src/sentry/grouping/parameterization.py

```diff
@@ -0,0 +1,368 @@
+
+        self.matches_counter: defaultdict[str, int] = defaultdict(int)
+
+    @staticmethod
+    def _make_regex_from_patterns(pattern_keys: Sequence[str]) -> re.Pattern[str]:
+        """
+        Takes list of pattern keys and returns a compiled regex pattern that matches any of them.
+
+        @param pattern_keys: A list of keys to match in the _parameterization_regex_components dict.
+        @returns: A compiled regex pattern that matches any of the given keys.
+        @raises: KeyError on pattern key not in the _parameterization_regex_components dict
+
+        The `(?x)` tells the regex compiler to ignore comments and unescaped whitespace,
+        so we can use newlines and indentation for better legibility in patterns above.
+        """
+
+        return re.compile(
+            rf"(?x){'|'.join(DEFAULT_PARAMETERIZATION_REGEXES_MAP[k] for k in pattern_keys)}"
+        )
+
+    def parametrize_w_regex(self, content: str) -> str:
+        """
+        Replace all matches of the given regex in the content with a placeholder string.
+
+        @param content: The string to replace matches in.
+        @param parameterization_regex: The compiled regex pattern to match.
+        @param match_callback: An optional callback function to call with the key of the matched pattern.
+
+        @returns: The content with all matches replaced with placeholders.
+        """
+
+        def _handle_regex_match(match: re.Match[str]) -> str:
+            # Find the first (should be only) non-None match entry, and sub in the placeholder. For
+            # example, given the groupdict item `('hex', '0x40000015')`, this returns '<hex>' as a
+            # replacement for the original value in the string.
+            for key, value in match.groupdict().items():
+                if value is not None:
+                    self.matches_counter[key] += 1
+                    return f"<{key}>"
+            return ""
+
+        return self._parameterization_regex.sub(_handle_regex_match, content)
+
+    def parametrize_w_experiments(
+        self, content: str, should_run: Callable[[str], bool] = lambda _: True
+    ) -> str:
+        """
+        Apply all experiments to the content.
+
+        @param content: The string to apply experiments to.
+        @returns: The content with all experiments applied.
+        """
+
+        def _incr_counter(key: str, count: int) -> None:
+            self.matches_counter[key] += count
+
+        def _handle_regex_match(match: re.Match[str]) -> str:
+            # Find the first (should be only) non-None match entry, and sub in the placeholder. For
+            # example, given the groupdict item `('hex', '0x40000015')`, this returns '<hex>' as a
+            # replacement for the original value in the string.
+            for key, value in match.groupdict().items():
+                if value is not None:
+                    self.matches_counter[key] += 1
+                    return f"<{key}>"
+            return ""
```

Comment:
Just FYI some coverage is lacking
<img width="295" alt="image" src="https://github.com/getsentry/sentry/assets/44410/a630ba72-b572-4523-bbd7-eb978f7877d3">